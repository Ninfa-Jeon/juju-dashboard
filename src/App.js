import React from 'react';
import {BrowserRouter as Router, Route, Link} from 'react-router-dom';

import Nav from './components/Nav/Nav';

// All following components are placeholders and will be replaced with imports.
function Models() {
  return <h2>Models</h2>;
}

function Clouds() {
  return <h2>Clouds</h2>;
}

function Kubernetes() {
  return <h2>Kubernetes</h2>;
}

function Controllers() {
  return <h2>Controllers</h2>;
}

function Usage() {
  return <h2>Usage</h2>;
}

function Logs() {
  return <h2>Logs</h2>;
}

function App() {
  return (
    <Router>
      <Nav />
      <div className="p-strip">
        <div className="row">
          <div className="col-3">
            <div className="p-card">
              <ul className="p-list">
                <li className="p-list__item">
                  <Link to="/">Models</Link>
                </li>
                <li className="p-list__item">
                  <Link to="/clouds">Clouds</Link>
                </li>
                <li className="p-list__item">
                  <Link to="/kubernetes">Kubernetes</Link>
                </li>
                <li className="p-list__item">
                  <Link to="/controllers">Controllers</Link>
                </li>
                <li className="p-list__item">
                  <Link to="/usage">Usage</Link>
                </li>
                <li className="p-list__item">
                  <Link to="/logs">Logs</Link>
                </li>
              </ul>
            </div>
          </div>
          <div className="col-9">
            <Route path="/" exact component={Models} />
            <Route path="/clouds" exact component={Clouds} />
            <Route path="/kubernetes" exact component={Kubernetes} />
            <Route path="/controllers" exact component={Controllers} />
            <Route path="/usage" exact component={Usage} />
            <Route path="/logs" exact component={Logs} />
          </div>
        </div>
      </div>
    </Router>
  );
}

export default App;
