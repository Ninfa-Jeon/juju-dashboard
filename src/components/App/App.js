import React from "react";
import { useSelector } from "react-redux";
import { BrowserRouter as Router, Route } from "react-router-dom";
import { QueryParamProvider } from "use-query-params";
import ReactGA from "react-ga";

// Components
import ErrorBoundary from "components/ErrorBoundary/ErrorBoundary";

import { Routes } from "components/Routes/Routes";

import { getConfig } from "app/selectors";
import useLocalStorage from "hooks/useLocalStorage";

function App() {
  const isProduction = process.env.NODE_ENV === "production" ?? true;
  const { baseAppURL } = useSelector(getConfig);
  const [disableAnalytics] = useLocalStorage("disableAnalytics", false);
  if (isProduction && (!disableAnalytics || disableAnalytics === "false")) {
    ReactGA.initialize("UA-1018242-68");
    ReactGA.pageview(window.location.href.replace(window.location.origin, ""));
  }

  return (
    <Router basename={baseAppURL}>
      <ErrorBoundary>
        <QueryParamProvider ReactRouterRoute={Route}>
          <Routes />
        </QueryParamProvider>
      </ErrorBoundary>
    </Router>
  );
}

export default App;
