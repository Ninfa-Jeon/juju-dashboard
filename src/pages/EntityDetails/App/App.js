import { useMemo } from "react";
import { useParams, useHistory } from "react-router-dom";
import { useQueryParam, StringParam, withDefault } from "use-query-params";
import MainTable from "@canonical/react-components/dist/components/MainTable";

import ButtonGroup from "components/ButtonGroup/ButtonGroup";
import InfoPanel from "components/InfoPanel/InfoPanel";
import EntityInfo from "components/EntityInfo/EntityInfo";

import EntityDetails from "pages/EntityDetails/EntityDetails";

import useModelStatus from "hooks/useModelStatus";

import {
  extractRevisionNumber,
  generateStatusElement,
  filterModelStatusDataByApp,
} from "app/utils/utils";

import { generateMachineRows, generateUnitRows } from "tables/tableRows";

import { machineTableHeaders, unitTableHeaders } from "tables/tableHeaders";

export default function App() {
  const { modelName, userName, appName: entity } = useParams();
  // Get model status info
  const modelStatusData = useModelStatus();
  const history = useHistory();

  // Filter model status via selected entity
  const filteredModelStatusData = filterModelStatusDataByApp(
    modelStatusData,
    entity
  );

  const app = modelStatusData?.applications[entity];

  const machinesPanelRows = useMemo(
    () => generateMachineRows(filteredModelStatusData),
    [filteredModelStatusData]
  );

  const unitPanelRows = useMemo(
    () => generateUnitRows(filteredModelStatusData),
    [filteredModelStatusData]
  );

  const [tableView, setTableView] = useQueryParam(
    "tableview",
    withDefault(StringParam, "units")
  );

  const showConfig = () => {
    history.push(
      `/models/${userName}/${modelName}/?entity=${entity}&panel=config`
    );
  };

  const AppEntityData = {
    status: app.status?.status ? generateStatusElement(app.status.status) : "-",
    charm: app.charm,
    os: "Ubuntu",
    revision: extractRevisionNumber(app.charm) || "-",
    message: "-",
  };

  return (
    <EntityDetails>
      <div>
        <InfoPanel />
        {modelStatusData && (
          <>
            <div className="entity__actions">
              <button
                className="entity-details__config-button"
                onClick={showConfig}
              >
                <i className="p-icon--settings"></i>Configure
              </button>
            </div>
            <EntityInfo data={AppEntityData} />
          </>
        )}
      </div>
      <div className="entity-details__content">
        <ButtonGroup
          buttons={["units", "machines"]}
          activeButton={tableView}
          setActiveButton={setTableView}
        />
        <div className="entity-details__tables">
          {tableView === "units" && (
            <MainTable
              headers={unitTableHeaders}
              rows={unitPanelRows}
              className="entity-details__units p-main-table panel__table"
              sortable
              emptyStateMsg={"There are no units in this model"}
            />
          )}
          {tableView === "machines" && (
            <MainTable
              headers={machineTableHeaders}
              rows={machinesPanelRows}
              className="entity-details__machines p-main-table panel__table"
              sortable
              emptyStateMsg={"There are no machines in this model"}
            />
          )}
        </div>
      </div>
    </EntityDetails>
  );
}
