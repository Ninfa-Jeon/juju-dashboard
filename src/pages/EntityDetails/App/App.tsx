import { useEffect, useMemo, useRef, useState } from "react";
import { Formik, Field } from "formik";
import { useParams } from "react-router-dom";
import {
  useQueryParam,
  useQueryParams,
  StringParam,
  withDefault,
} from "use-query-params";
import MainTable from "@canonical/react-components/dist/components/MainTable";

import ButtonGroup from "components/ButtonGroup/ButtonGroup";
import InfoPanel from "components/InfoPanel/InfoPanel";
import EntityInfo from "components/EntityInfo/EntityInfo";
import ChipGroup from "components/ChipGroup/ChipGroup";
import FormikFormData from "components/FormikFormData/FormikFormData";

import EntityDetails from "pages/EntityDetails/EntityDetails";

import useModelStatus from "hooks/useModelStatus";
import useTableRowClick from "hooks/useTableRowClick";

import {
  extractRevisionNumber,
  generateStatusElement,
  filterModelStatusDataByApp,
} from "app/utils/utils";

import type { EntityDetailsRoute } from "components/Routes/Routes";
import type { SetFieldValue } from "components/FormikFormData/FormikFormData";
import type { TSFixMe } from "types";

import { generateMachineRows, generateUnitRows } from "tables/tableRows";
import {
  machineTableHeaders,
  generateSelectableUnitTableHeaders,
} from "tables/tableHeaders";

import { renderCounts } from "../counts";

type FormData = {
  selectAll: boolean;
  selectedUnits: string[];
};

export default function App(): JSX.Element {
  const { appName: entity } = useParams<EntityDetailsRoute>();
  const [selectedUnits, setSelectedUnits] = useState<string[]>([]);
  const [selectAll, setSelectAll] = useState<boolean>(false);

  const tablesRef = useRef<HTMLDivElement>(null);
  const setFieldsValues = useRef<SetFieldValue>();
  // Get model status info
  const modelStatusData: TSFixMe = useModelStatus();

  const tableRowClick = useTableRowClick();

  // Filter model status via selected entity
  const filteredModelStatusData = filterModelStatusDataByApp(
    modelStatusData,
    entity
  );

  const app = modelStatusData?.applications[entity];

  const machinesPanelRows = useMemo(
    () => generateMachineRows(filteredModelStatusData, tableRowClick),
    [filteredModelStatusData, tableRowClick]
  );

  const unitTableHeaders = useMemo(() => {
    const fieldID = "unit-list-select-all";
    return generateSelectableUnitTableHeaders({
      content: (
        <label className="p-checkbox" htmlFor={fieldID}>
          <Field
            id={fieldID}
            type="checkbox"
            aria-labelledby="checkboxLabel0"
            className="p-checkbox__input"
            name="selectAll"
          />
          <span className="p-checkbox__label" id="checkboxLabel0"></span>
        </label>
      ),
      sortKey: "",
      className: "select-unit",
    });
  }, []);

  const unitPanelRows = useMemo(
    () => generateUnitRows(filteredModelStatusData, tableRowClick, true),
    [filteredModelStatusData, tableRowClick]
  );

  const [tableView, setTableView] = useQueryParam(
    "tableview",
    withDefault(StringParam, "units")
  );

  const [query, setQuery] = useQueryParams({
    panel: StringParam,
    entity: StringParam,
    activeView: withDefault(StringParam, "apps"),
  });

  const showConfig = () => {
    query && setQuery({ panel: "config", entity: entity });
  };

  const AppEntityData = {
    status:
      app && app.status?.status
        ? generateStatusElement(app.status.status)
        : "-",
    charm: app?.charm,
    os: "Ubuntu",
    revision: (app && extractRevisionNumber(app.charm)) || "-",
    message: "-",
  };

  const unitChips = renderCounts("units", modelStatusData);
  const machineChips = renderCounts("machines", modelStatusData);

  const setPanel = useQueryParam("panel", StringParam)[1];
  const showActions = () => {
    setPanel("execute-action");
  };

  useEffect(() => {
    if (setFieldsValues.current) {
      if (selectAll) {
        setFieldsValues.current("selectedUnits", Object.keys(app.units));
      }
    }
  }, [selectAll, app]);

  useEffect(() => {
    if (setFieldsValues.current) {
      const unitCount = app?.units ? Object.keys(app.units).length : null;
      // If all of the units are selected then setSelectAll true
      if (selectedUnits.length === unitCount && !selectAll) {
        setFieldsValues.current("selectAll", true);
      }
      // If any of the units aren't selected then setSelectAll false
      if (selectedUnits.length !== unitCount && selectAll) {
        setFieldsValues.current("selectAll", false);
      }
    }
  }, [selectAll, selectedUnits, app]);

  const onFormChange = (formData: FormData) => {
    setSelectAll(formData.selectAll);
    setSelectedUnits(formData.selectedUnits);
  };

  const onSetup = (setFieldValue: SetFieldValue) => {
    setFieldsValues.current = setFieldValue;
  };

  return (
    <EntityDetails type="app" className="entity-details__app">
      <div>
        <InfoPanel />
        <>
          <div className="entity-details__actions">
            <button
              className="entity-details__action-button"
              onClick={showConfig}
            >
              <i className="p-icon--settings"></i>Configure
            </button>

            <button
              className="entity-details__action-button"
              onClick={showActions}
            >
              <i className="p-icon--settings"></i>Actions
            </button>
          </div>
          <EntityInfo data={AppEntityData} />
        </>
      </div>
      <div className="entity-details__main u-overflow--scroll">
        <ButtonGroup
          buttons={["units", "machines"]}
          activeButton={tableView}
          setActiveButton={setTableView}
        />
        <div className="entity-details__tables" ref={tablesRef}>
          {tableView === "units" && (
            <>
              <ChipGroup chips={unitChips} descriptor="units" />
              <Formik
                initialValues={{
                  selectAll: false,
                  selectedUnits: [],
                }}
                onSubmit={() => {}}
              >
                <FormikFormData onFormChange={onFormChange} onSetup={onSetup}>
                  <MainTable
                    headers={unitTableHeaders}
                    rows={unitPanelRows}
                    className="entity-details__units p-main-table panel__table"
                    sortable
                    emptyStateMsg={"There are no units in this model"}
                  />
                </FormikFormData>
              </Formik>
            </>
          )}
          {tableView === "machines" && (
            <>
              <ChipGroup chips={machineChips} descriptor="machines" />
              <MainTable
                headers={machineTableHeaders}
                rows={machinesPanelRows}
                className="entity-details__machines p-main-table panel__table"
                sortable
                emptyStateMsg={"There are no machines in this model"}
              />
            </>
          )}
        </div>
      </div>
    </EntityDetails>
  );
}
