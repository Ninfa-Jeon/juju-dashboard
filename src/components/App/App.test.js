import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { mount } from "enzyme";
import App from "./App";

import dataDump from "../../testing/complete-redux-store-dump";

const mockStore = configureStore([]);

describe("App", () => {
  it("properly sets up Router", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(wrapper.find("BrowserRouter").prop("basename")).toBe("/");
  });
});
