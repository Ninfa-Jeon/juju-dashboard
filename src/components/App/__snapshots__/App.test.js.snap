// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders without crashing and matches snapshot 1`] = `
<BrowserRouter>
  <ErrorBoundary>
    <Switch>
      <Route
        component={[Function]}
        exact={true}
        path="/"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/models/:id"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/clouds"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/kubernetes"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/controllers"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/usage"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/logs"
      />
      <Route
        component={[Function]}
      />
    </Switch>
  </ErrorBoundary>
</BrowserRouter>
`;
